homeassistant/components/recorder/__init__.py

```diff
@@ -788,34 +802,56 @@ def _run_external_statistics(self, metadata, stats):
         # Schedule a new statistics task if this one didn't finish
         self.queue.put(ExternalStatisticsTask(metadata, stats))
 
-    def _process_one_event(self, event):
+    def _lock_database(self, task: DatabaseLockTask):
+        with write_lock_db(self):
+            # Notify that lock is being held, wait until database can be used again.
+            task.database_locked.set()
+            while not task.database_unlock.wait(timeout=1):
+                if self.queue.qsize() > MAX_QUEUE_BACKLOG * 0.9:
+                    _LOGGER.warning(
+                        "Database queue backlog reached more than 90% of maximum queue length. Continue writing. Your Backup might be corrupted."
+                    )
+                    task.queue_overflow = True
+                    break
+        _LOGGER.info(
+            "Database queue backlog reached %d entries during backup.",
+            self.queue.qsize(),
+        )
+
+    def _process_one_task(self, event) -> bool:
```

Comment:
For a future PR, this linear `isinstance` search should probably be a `dict` lookup

Reply:
Yeah or using a Task wrapper class for `Event` too, and introduce a base class with a handler funciton or similar. The reason I split it in this PR was a linter error. Agreed, optimizing it should be in a future PR.